exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -869,139 +872,99 @@ func (b *Bitfinex) handleWSTickerUpdate(c *subscription.Subscription, d []interf
-			} else {
-				price, ok := elem[3].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade price")
-				}
-				wsTrade.Rate = price
-			}
-			tradeHolder = append(tradeHolder, wsTrade)
-		}
-	case 3:
-		if eventType != wsFundingTradeUpdate && eventType != wsTradeExecutionUpdate {
-			return fmt.Errorf("unhandled WS trade update event: %s", eventType)
+	var wsTrades []*wsTrade
+	switch valueType {
+	case jsonparser.String:
+		if t, err := b.handleWSPublicTradeUpdate(respRaw); err != nil {
+			return err
+		} else {
+			wsTrades = []*wsTrade{t}
 		}
-		data, ok := d[2].([]interface{})
-		if !ok {
-			return errors.New("trade data type assertion error")
+	case jsonparser.Array:
+		if wsTrades, err = b.handleWSPublicTradesSnapshot(respRaw); err != nil {
+			return err
 		}
-
-		tradeID, ok := data[0].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade ID")
+	default:
+		return fmt.Errorf("%w `tradesUpdate[1]`: %w `%s`", errParsingWSField, jsonparser.UnknownValueTypeError, valueType)
+	}
+	trades := make([]trade.Data, len(wsTrades))
+	for i, w := range wsTrades {
+		t := trade.Data{
+			Exchange:     b.Name,
+			AssetType:    s.Asset,
+			CurrencyPair: s.Pairs[0],
+			TID:          strconv.FormatInt(w.ID, 10),
+			Timestamp:    w.Timestamp.Time().UTC(),
+			Side:         order.Buy,
+			Amount:       w.Amount,
+			Price:        w.Price,
 		}
-		timestamp, ok := data[1].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade timestamp")
+		if w.Period != 0 {
+			t.AssetType = asset.MarginFunding
 		}
-		amount, ok := data[2].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade amount")
+		if t.Amount < 0 {
+			t.Side = order.Sell
+			t.Amount *= -1
 		}
-		wsTrade := WebsocketTrade{
-			ID:        int64(tradeID),
-			Timestamp: int64(timestamp),
-			Amount:    amount,
+		if feedEnabled {
+			b.Websocket.DataHandler <- &t
 		}
-		if len(data) == 5 {
-			rate, ok := data[3].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade rate")
-			}
-			period, ok := data[4].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade period")
-			}
-			wsTrade.Rate = rate
-			wsTrade.Period = int64(period)
+		trades[i] = t
+	}
+	if b.IsSaveTradeDataEnabled() {
+		err = trade.AddTradesToBuffer(b.GetName(), trades...)
+	}
+	return err
+}
+
+func (b *Bitfinex) handleWSPublicTradesSnapshot(respRaw []byte) (trades []*wsTrade, errs error) {
+	handleTrade := func(v []byte, valueType jsonparser.ValueType, _ int, _ error) {
+		if valueType != jsonparser.Array {
+			errs = common.AppendError(errs, fmt.Errorf("%w `tradesSnapshot[1][*]`: %w `%s`", errParsingWSField, jsonparser.UnknownValueTypeError, valueType))
 		} else {
-			price, ok := data[3].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade price")
+			t := &wsTrade{}
+			if err := json.Unmarshal(v, &[]any{&t.ID, &t.Timestamp, &t.Amount, &t.Price, &t.Period}); err != nil {
```

Comment:
Suggestion: can define an UnmarshalJSON method with this and use the type `wsTrade` here and below in `handleWSPublicTradeUpdate`.

Reply:
Oh, yeah, that's really cute, thank you 🙏 
I've put it in `_types` since it's a micro-method and there's nowhere moar organic to put it, other than maybe in `_websocket` next to the funcs that'll use it.

Fixed https://github.com/gbjk/gocryptotrader/commit/4d3a7e0a4